File: go/minstack.go
package main

import "fmt"

type MinStack struct {
	stack []int
	mins  []int
}

func NewMinStack() *MinStack {
	return &MinStack{make([]int, 0), make([]int, 0)}
}

func (st *MinStack) String() string {
	if st.Empty() {
		return "empty stack"
	}
	return fmt.Sprintf("top=%d min=%d", st.Top(), st.Min())
}

func (st *MinStack) Empty() bool {
	return len(st.stack) == 0
}

func (st *MinStack) Push(n int) {
	st.stack = append(st.stack, n)
	if len(st.mins) == 0 || n <= st.mins[len(st.mins)-1] {
		st.mins = append(st.mins, n)
	}
}

func (st *MinStack) Top() int {
	if st.Empty() {
		panic("empty stack")
	}
	return st.stack[len(st.stack)-1]
}

func (st *MinStack) Pop() int {
	if st.Empty() {
		panic("empty stack")
	}
	n := st.stack[len(st.stack)-1]
	st.stack = st.stack[0 : len(st.stack)-1]
	if n == st.mins[len(st.mins)-1] {
		st.mins = st.mins[0 : len(st.mins)-1]
	}
	return n
}

func (st *MinStack) Min() int {
	if st.Empty() {
		panic("empty stack")
	}
	return st.mins[len(st.mins)-1]
}

func main() {
	st := NewMinStack()
	st.Push(3)
	st.Push(4)
	st.Push(5)
	fmt.Println(st)
	st.Push(1)
	st.Push(2)
	fmt.Println(st)
	st.Pop()
	st.Pop()
	fmt.Println(st)
}

File: python/src/min_stack.py
# See http://programmingpraxis.com/2012/07/27/min-stack/

import unittest


class MinStack(object):

    def __init__(self):
        self.stack = []
        self.mins = []

    def push(self, item):
        self.stack.append(item)
        if not self.mins or item <= self.mins[-1]:
            self.mins.append(item)

    def pop(self):
        item = self.stack.pop()
        if item == self.mins[-1]:
            self.mins.pop()
        return item

    def min(self):
        return self.mins[-1]

    def __len__(self):
        return len(self.stack)


class Test(unittest.TestCase):

    def test_empty(self):
        ms = MinStack()
        self.assertEqual(len(ms), 0)

    def test_push(self):
        ms = MinStack()
        ms.push(1)
        ms.push(2)
        self.assertEqual(len(ms), 2)

    def test_pop(self):
        ms = MinStack()
        ms.push(1)
        ms.push(2)
        self.assertEqual(ms.pop(), 2)
        self.assertEqual(ms.pop(), 1)
        self.assertEqual(len(ms), 0)

    def test_min(self):
        ms = MinStack()
        ms.push(2)
        ms.push(3)
        self.assertEqual(ms.min(), 2)
        ms.push(0)
        ms.push(1)
        self.assertEqual(ms.min(), 0)
        ms.push(-1)
        ms.push(4)
        self.assertEqual(ms.min(), -1)
        ms.pop()
        ms.pop()
        self.assertEqual(ms.min(), 0)


if __name__ == '__main__':
    unittest.main()
